### PR TITLE
Add the `balance_sheet_assets_ferc1` table

### PR DESCRIPTION
**Pull request by cmgosnell**
# Overview
- Added a new table
- It needed both a `wide_to_tidy` and a `process_instant_xbrl` that pulls the `starting_balance` and `ending_balance` into the same record.
- One oddity of this table is that the DBF data has multiple rows where the same data is reported. Believe it or not, the data in there IS THE SAME! I'm very surprised. In the `process_dbf` step, I added a new transformer called `drop_duplicate_rows_dbf` which checks whether or not the duplicated records do indeed contain the same values, then drop one set of dupes.

## PR Checklist

Before requesting a review of your pull request, please make sure you've done the
following:

* [x] Merge the most recent version of `dev` (or the appropriate upstream branch) into
      your branch and resolved any merge conflicts. You may need to do this several
      times over the course of a PR as `dev` changes frequently.
* [x] Verify that all of the CI checks on your PR are passing. See
      [Running Tests with Tox](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
      for details on how to run the full test suite locally if you need to debug a
      particular failure.
* [x] Ensure that the docstrings for any new modules, classes, functions, or methods are
      descriptive enough for developers and users to understand your code.
* [x] If you've added new functions or classes, ensure that they have at least basic
      unit tests.
* [x] Do your own review of the PR. Add comments highlighting areas where you have
      questions you'd like reviewers to answer, known issues, solutions you're
      unsatisfied with, or other things that deserve special attention from the
      reviewer.
* [ ] Update the
      [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html)
      to reflect your changes. Make sure to reference the PR and any related issues.
* [ ] If you expanded data coverage or changed the outputs, ensure that the full
      [data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
      pass locally on a fresh DB.
* [ ] If you've added new analyses, make sure they include defensive sanity checks that
      will catch unexpected data issues.

### CODE REVIEW

File: src/pudl/transform/ferc1.py [hunk 1]
```diff
@@ -2759,6 +2820,118 @@ def normalize_metadata_xbrl(
...
+    def targeted_drop_duplicates_dbf(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Drop bad duplicate records from a specific utility in 1994.
+
+        This is a very specific fix, meant to get rid of a particular observed set of
+        duplicate records: FERC Respondent ID 165 in 1999 has two sets of assets and
+        other debt records, one of which contains one null record.
+
+        This method is part of the DBF processing because we want to be able to
+        hard-code a specific value of ``utility_id_ferc1_dbf`` and those IDs are no
+        longer available later in the process. I think.
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:2838]: ah sorry i copied this from one of zane's `targeted_drop...` methods. I will clean this up.

File: src/pudl/transform/ferc1.py [hunk 2]
```diff
@@ -2759,6 +2820,118 @@ def normalize_metadata_xbrl(
...
+    table_id: TableIdFerc1 = TableIdFerc1.BALANCE_SHEET_ASSETS_FERC1
+    has_unique_record_ids: bool = False
+
+    def targeted_drop_duplicates_dbf(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Drop bad duplicate records from a specific utility in 1994.
+
+        This is a very specific fix, meant to get rid of a particular observed set of
+        duplicate records: FERC Respondent ID 165 in 1999 has two sets of assets and
+        other debt records, one of which contains one null record.
+
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:2835]: hm maybe we want to make a more generic `target_drop` transformer... I could certainly imagine doing this but I think it would be hard to make it general enough and useful enough.

A lower hanging fruit option would be to make a no-op version of `targeted_drop_duplicates_dbf` within `Ferc1AbstractTableTransformer` that just takes a df and returns it within `process_dbf`. This way any table-transformer could override it with any bespoke needs and we could incorporate the three steps I added to this table's `process_dbf` to the generic version (I added `drop_invalid_rows`, `targeted_drop_duplicates_dbf` and `drop_duplicate_rows_dbf`).

We coooould also incorporate the dropping of records that *are* PK duplicates but have one (or more) null data value into `drop_duplicate_rows_dbf` to avoid the addition of `drop_invalid_rows` and `targeted_drop_duplicates_dbf`. I think I am partial to this, especially if you say your found this issue in your table as well.

File: src/pudl/transform/ferc1.py [hunk 3]
```diff
@@ -222,6 +223,53 @@ def merge_metadata_xbrl(
...
+            .schema.primary_key
+        )
+        # checks to make sure the drop is targeted as expected
+        pk_dupes = df[df.duplicated(pks, keep=False)]
+        pk_data_dupes = pk_dupes[
+            pk_dupes.duplicated(pks + params.data_columns, keep=False)
+        ]
+        dupes_w_unique_data = pk_dupes.loc[
+            pk_dupes.index.difference(pk_data_dupes.index)
+        ]
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:260]: ah ty!! that felt overly complex!

File: src/pudl/transform/ferc1.py [hunk 4]
```diff
@@ -2759,6 +2820,118 @@ def normalize_metadata_xbrl(
...
+        meta = (
+            super()
+            .normalize_metadata_xbrl(xbrl_fact_names)
+            .assign(
+                xbrl_factoid=lambda x: x.xbrl_fact_name,
+            )
+        )
+        # Save the normalized metadata so it can be used by other methods.
+        self.xbrl_metadata_normalized = meta
+        return meta
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:2895]: 100%! I think @zaneselvans is handling this as he is integrating the new metadata.

I believe the outcome of it will be removing the need for all of these direct copy/paste methods that are just adding a column with a different name.

File: src/pudl/transform/ferc1.py [hunk 5]
```diff
@@ -2759,6 +2820,118 @@ def normalize_metadata_xbrl(
...
+
+    @cache_df("process_instant_xbrl")
+    def process_instant_xbrl(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Pre-processing required to make the instant and duration tables compatible.
+
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:2900]: Yes! If this works for your table, my table and the plant-in-service table... I'm tempted to go ahead and make a transformer for it.

[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/transform/ferc1.py:2900]: Yes! As long as it has the caveat I included in my PR for years (or something similar?)

File: src/pudl/transform/params/ferc1.py [hunk 6]
```diff
@@ -3213,6 +3213,57 @@
...
+        },
+        "align_row_numbers_dbf": {"dbf_table_name": "f1_comp_balance_db"},
+        "merge_metadata_xbrl": {
+            "rename_columns": {"xbrl_factoid": "asset_type"},
+            "on": "asset_type",
+        },
+        "drop_invalid_rows": [
+            {
+                "invalid_values": [pd.NA, np.nan, ""],
+                "required_valid_cols": ["starting_balance", "ending_balance"],
```
[the remark this answers is not in the document]
Comment (aesharpe) [reply src/pudl/transform/params/ferc1.py:3263]: Maybe we should ask RMI about this?

File: src/pudl/transform/ferc1.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:2900]: we discussed this and @aesharpe is planning on taking a stab at the normalization of this in #2134

File: src/pudl/transform/ferc1.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py]: updated and then deleted entirely! bc this functionality got squished into `drop_duplicate_rows_dbf`

File: src/pudl/transform/ferc1.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py]: we went with this last suggestion: update `drop_duplicate_rows_dbf` so it drops dupes with *exactly* the same data as well as records with one null-having row and one data-full row.

File: src/pudl/transform/ferc1.py [hunk 7]
```diff
@@ -222,6 +223,53 @@ def merge_metadata_xbrl(
...
+
+def drop_duplicate_rows_dbf(df, params: DropDuplicateRowsDbf):
+    """Drop the duplicate DBF rows if all duplicates have indentical data.
+
+    There are several instances of the DBF data reporting the same value on multiple
+    rows. This function checks to see if all of the duplicate values that have the same
+    primary keys have reported the same data. If the duplicates have the same data, the
+    duplicates are dropped with ``keep="first"``. If any duplicates do not contain the
+    same data, an assertion will be raised.
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:243]: after discussing this we decided to update `drop_duplicate_rows_dbf` so it drops dupes with exactly the same data as well as records with one null-having row and one data-full row.

File: src/pudl/transform/ferc1.py [hunk 8]
```diff
@@ -242,31 +242,42 @@ def drop_duplicate_rows_dbf(df, params: DropDuplicateRowsDbf):
...
-            f"Dropping {(len(pk_data_dupes)/2)/len(df):.1%} of duplicate records."
         )
-        df = df.drop_duplicates(pks, keep="first")
+    )
+    if not dupes_w_unique_data[
+        dupes_w_unique_data.count_of_null_vs_non_null != 2
+    ].empty:
+        raise AssertionError(
+            "Duplicates have unique data and should not be dropped. Unique data: "
+            f"{len(dupes_w_unique_data)}: {dupes_w_unique_data}"
```
Comment (aesharpe) [inline src/pudl/transform/ferc1.py:273]: This seems good and will work for the case in my #2134 PR.

However, there are some edge cases that would bypass this assertion error:
- dupes in `dupes_w_unique_data` that both have some NA values (this seems tricky any maybe something we should combine? But highly uncertain about this. Depends on the case.)
- dupes in `dupes_w_unique_data` where one row has some NA values but other non-NA values that don't match the other duplicate (idk what we would even do with this besides flag it...)

File: src/pudl/transform/ferc1.py [hunk 9]
```diff
@@ -242,31 +242,42 @@ def drop_duplicate_rows_dbf(df, params: DropDuplicateRowsDbf):
...
+        )
+    len_og = len(df)
+    df = (
+        df.sort_values(by=["null_data"], ascending=True)
+        .drop_duplicates(pks, keep="first")
+        .drop(columns=["null_data"])
+    )
```
Comment (aesharpe) [inline src/pudl/transform/ferc1.py:280]: This seems like a good way to make sure to keep the non null dups (as long as ascending=True puts False above True!)

Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:280]: it does! I always have to double check and I did add a case into the unit tests to ensure this

File: src/pudl/transform/ferc1.py (hunk 8 above)
Comment (cmgosnell) [reply src/pudl/transform/ferc1.py:273]: I'm glad it will probably work with your table!

I'm glad you are noting these edge cases. I believe the first example that you noted will actually raise the assertion. And I think that is a good thing for now. I could definitely imagine developing a transformer or a part of this transformer that would condense these, but I would suggest we cross that bridge when we get there. For now, I added a small unit test  to be sure that these types of records trip the assertion:
```
# if a PK dupes has mismatched null data
df_nulls = pd.read_csv(
    StringIO(
        """
report_year,utility_id_ferc1,asset_type,data_col1,data_col2
2021,71,stuff,,700
2021,71,stuff,70,
"""
    )
)
with pytest.raises(AssertionError):
    drop_duplicate_rows_dbf(df_nulls, params=params)
```

you are very right that this would not raise an error...
```
report_year,utility_id_ferc1,asset_type,data_col1,data_col2
2021,71,stuff,,701
2021,71,stuff,70,700
```
I can definitely imagine this showing up. sigh. I'll try to weasel in a way to make this raise the assertion and add it as a unit test.

Comment (cmgosnell) [reply src/pudl/transform/ferc1.py]: okay I edited the assertion in there and added a section to the unit test to catch this second edge case

Comment (aesharpe) [reply src/pudl/transform/ferc1.py]: It takes me like 5 min every time to wrap my head around all the negatives in 

```
if not dupes_w_unique_data[
        dupes_w_unique_data.count_of_null_vs_non_null != 2
    ].empty
```
lolol

**Review by aesharpe — APPROVED** [review]
